### PR TITLE
Add command-line db-manager

### PR DESCRIPTION
**Pull request by xmo-odoo**
Currently dbs can only be managed via the UI in order to take filestores in account: while it's possible to load/copy/rename/drop databases via `psql`, that will not manage the related filestores so the result of the operation is incomplete DBs and leftover filestores littering the disk.

Seems like a good idea to add a CLI to perform the same tasks. Currently the CLI calls into the corresponding service, rather than both calling into (possibly better designed) unified APIs, but that seems fine for an initial version.

The top-level `db` command acts as a db manager, with git-style sub-sub-commands for the various operations:

- `load` to load a dump file into a database (with a specified name or not)
- `dump` to dump a local db to a zip dump (pg_dump can be created via the corresponding command so not a concern)
- `duplicate` and `rename`
- `drop` in order to drop both the database itself and the corresponding filestore

Notably, `create` is currently left out because a database can trivially be created by invoking odoo using a dbname which doesn't exist, so doesn't seem useful.

`list` is also left out, because `psql -l` generally does the job.

### CODE REVIEW

**Comment by xmo-odoo** [comment]
Missing option: `-D` / `--data-dir` option (that's where the filestore actually goes)

Maybe to add: allow an URL and go fetch the dump file over the internets? Way more convenient than downloading manually.

**Comment by xmo-odoo** [comment]
Also maybe put the "regular odoo" options in a group of their own for a cleaner CLI?

**Comment by Yenthe666** [comment]
> Should load_db automatically run neutralize, or have an option to do so? I don't think running the two subcommands one after the other is very difficult but...
I personally think that having an option to add this flag would be great :)

File: odoo/cli/load_db.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+        too) their regular Odoo behaviour.
+        """
+        parser = argparse.ArgumentParser(
+            prog="%s %s" % (Path(argv[0]).name, self.name),
+            description=self.__doc__
+        )
+        parser.add_argument(
+            '-d', '--database',
+            help="Database to restore to, if not provided uses the dump file's"
+                 " filename (without extension"
```
Comment (Julien00859) [inline odoo/cli/load_db.py:27]:
```suggestion
            help="Database to restore to, if not provided uses the dump file's"
                 " filename (without extension)"
```

File: odoo/cli/load_db.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        parser.add_argument('--pg_path')
+        parser.add_argument('--db_host')
+        parser.add_argument('--db_port')
+        parser.add_argument('--db_sslmode')
+        parser.add_argument('dump_file')
+        args = parser.parse_args(cmdargs)
+
+        # standard options parser doesn't like positional arguments, so strip
+        # the one
+        cmdargs.remove(args.dump_file)
```
Comment (Julien00859) [inline odoo/cli/load_db.py:42]: I don't like this `cmdargs.remove`

    odoo-bin load_db -d foo -c config.cfg foo

becomes

   odoo-bin load_db -d -c config.cfg foo

**Comment by xmo-odoo** [comment]
Changes:

- add support for downloading dump files from the internet
- add some progress output
- serialize Odoo options back from what we parsed (excluding `dump_file`) instead of trying to adjust the existing `cmdargs`

File: odoo/cli/load_db.py [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+        parser.add_argument('dump_file')
+        args = parser.parse_args(cmdargs)
+
+        config.parse_config([
+            val
+            for k, v in vars(args).items()
+            if v is not None
+            if k != 'dump_file'
+            for val in [k, v]
+        ])
```
Comment (Julien00859) [inline odoo/cli/load_db.py:51]: So this bit is to reconstruct a sys.argv like list without the positional `dump_file` am I right?

So if we do `odoo-bin load_db --db_host localhost -c config.cfg path/to/my/dump.sql` doesnt' this list comprehension gives `['db_host', 'localhost', 'config', 'config.cfg']`, aren't the dashes missing for db_host and config?

Shouldn't it be `for var in [f'--{k}', v]`?

Comment (xmo-odoo) [reply odoo/cli/load_db.py:51]: > Shouldn't it be `for var in [f'--{k}', v]`?

Yep, kinda forgot to test that... also needs to revert the key back to the CLI form, as argparse will convert the dashes to underscores...

Comment (Julien00859) [reply odoo/cli/load_db.py:51]: For what it's worth

    parser = ArgumentParser()
    next((action.option_strings[0] for action in parser._actions if action.dest == 'help'))
    '-h'

Comment (xmo-odoo) [reply odoo/cli/load_db.py:51]: Yeah I know. But I can probably just have a map for the exceptions and return the others as-is.

**Comment by xmo-odoo** [comment]
@Xavier-Do new version after discussion: for coherence's sake, proposal goes from just loading dumps to a "full-blown" db-manager on the command-line.

It doesn't do the things which can be done via `psql` directly, by design. Because you can use psql.

**Comment by xmo-odoo** [comment]
@robodoo override=ci/security

File: odoo/service/db.py [hunk 4]
```diff
@@ -275,13 +275,13 @@ def chunks(d, n=8192):
         os.unlink(data_file.name)
     return True
 
-@check_db_management_enabled
```
Comment (Xavier-Do) [inline odoo/service/db.py:278]: Why did you remove this?

File: odoo/cli/db.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    def drop(self, args):
+        if not exp_drop(args.database):
+            exit(f"Database {args.database} does not exist.")
+
+    def _check_target(self, target, *, delete_if_exists):
+        if exp_db_exist(target):
+            if delete_if_exists:
+                exp_drop(target)
+            else:
+                exit(f"Target database {target} exists, aborting.")
```
Comment (Xavier-Do) [inline odoo/cli/db.py:152]: We could simplify this by removing the delete_if_exists parameters using args.force

This will also allow to says "use --force to drop the existing database" in the message in order to help the user.

File: odoo/cli/db.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+        restore_db(
+            db=args.database or Path(args.dump_file).stem,
+            dump_file=dump_file,
+            copy=True
+        )
+
+    def dump(self, args):
+        if args.dump_path == '-':
+            dump_db(args.database, stdout.buffer)
```
Comment (Xavier-Do) [inline odoo/cli/db.py:130]: Nice 💯

File: odoo/service/db.py (hunk 4 above)
Comment (xmo-odoo) [reply odoo/service/db.py:278]: Because it seemed useless initially and I hadn't overridden `list_db`. Good spot.

If I remove any it's probably the one on `exp_restore` which is redundant as the http endpoint calls restore_db directly. Same for `dump_db`.

File: odoo/cli/db.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+            ]
+        ])
+        # force db management active to bypass check when only a
+        # `check_db_management_enabled` version is available.
+        config['list_db'] = True
+        report_configuration()
+
+        args.func(args)
+
+    def load(self, args):
```
Comment (Xavier-Do) [inline odoo/cli/db.py:108]: would be great to add a _check_target here (as well as the command line argument)

File: odoo/cli/db.py [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+        url = urllib.parse.urlparse(args.dump_file)
+        if url.scheme:
+            stderr.write(f"Fetching {args.dump_file}...")
+            r = requests.get(args.dump_file, timeout=10, stream=True)
```
Comment (Xavier-Do) [inline odoo/cli/db.py:112]: I don't really understand how it would work testing it i have 
```
TypeError: expected str, bytes or os.PathLike object, not HTTPResponse
```

**Review by Xavier-Do — CHANGES_REQUESTED** [review]

File: odoo/service/db.py (hunk 4 above)
Comment (Xavier-Do) [reply odoo/service/db.py]: I'm not sure it is useful but better safe than sorry :) With the config hack it should work fine.

File: odoo/cli/db.py (hunk 5 above)
Comment (xmo-odoo) [reply odoo/cli/db.py:152]: > We could simplify this by removing the delete_if_exists parameters using args.force

That would require passing the entire `args` to the helper and I don't want to do that because it makes the interface shit.

Comment (Xavier-Do) [reply odoo/cli/db.py:152]: Making the args global (or an attribute) would solve this issue by removing args parameter everywhere.

File: odoo/cli/db.py (hunk 8 above)
Comment (xmo-odoo) [reply odoo/cli/db.py:112]: Hmmm guessing the original put everything in memory wrapped in a `BytesIO`, but then I figured the response was file-like so I could just pass it directly to `restore_db`, and then I forgot to test that, and it doesn't work because the http stream is not seekable.

File: odoo/cli/db.py (hunk 5 above)
Comment (xmo-odoo) [reply odoo/cli/db.py:152]: Making the interface even worse is not what I had in mind.

File: odoo/cli/db.py (hunk 8 above)
Comment (Xavier-Do) [reply odoo/cli/db.py]: http stream should be seekable but it is passed directly to zipfile.zipfile and it's expecting a filename.

Comment (xmo-odoo) [reply odoo/cli/db.py]: > http stream should be seekable

It's not, I checked. Even passing it directly to `ZipFile` blows up,

>  but it is passed directly to zipfile.is_zipfile and it's expecting a filename.

Nope, takes a file-like object or a filename, but apparently if it gets a non-seekable file-like object it just returns `False` rather than an error, so we get the second path in `restore_db`.

But that means I should probably `is_zipfile` before calling `restore_db` otherwise it treats whatever nonsense we got as a raw postgres dump file and tries to `pg_restore` it. Which is not really necessary on the command line because you can just `pg_restore` it yourself like a big boy.

> My tests command line: `./odoo-bin db load http://runbot57.odoo.com/runbot/static/build/18642143-master/logs/18642143-master-all.zip`

Yeah that's what I checked with (and pretty much the entire point of the feature). I reversed it.

File: odoo/cli/db.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+            dump_file = io.BytesIO(r.content)
+        else:
+            stderr.write(f"Restoring {args.dump_file}...\n")
+            dump_file = args.dump_file
+
+        if not zipfile.is_zipfile(dump_file):
+            exit("Not a zipped dump file, use `pg_restore` to restore raw dumps,"
+                 " and `psql` to execute sql dumps or scripts.")
+
+        restore_db(db=db_name, dump_file=dump_file, copy=True)
```
Comment (Xavier-Do) [inline odoo/cli/db.py:135]: This part is a little fragile because of the loading of the registry. We may have noise or even crash trying
- to download a enterprise database without giving the correct addons_path
- to download a 13.0 database (will crash)

The copy option may be needed but not always, we may want to add an option for that.

Finding the filestore using the registry looks useless (we can check but I can't find why this is needed... )

We may change slightly the restore_db to find the filestore using the config  and not load the registry if we don't have the copy option. This would allow us to use this instead of the odooget script.

File: odoo/cli/db.py [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+
+        args.func(args)
+
+    def load(self, args):
+        db_name = args.database or Path(args.dump_file).stem
+        self._check_target(db_name, delete_if_exists=args.force)
+
+        url = urllib.parse.urlparse(args.dump_file)
+        if url.scheme:
+            stderr.write(f"Fetching {args.dump_file}...")
```
[the remark this answers is not in the document]
Comment (xmo-odoo) [reply odoo/cli/db.py:120]: That seems odd, stderr should be unbuffered...

[the remark this answers is not in the document]
Comment (xmo-odoo) [reply odoo/cli/db.py:120]: WTF and TIL: apparently python buffers stderr...

**Comment by xmo-odoo** [comment]
@robodoo r+

**Comment by xmo-odoo** [comment]
@robodoo r+
